### PR TITLE
Integration with Google Tag Manager

### PR DESCRIPTION
**Pull request by alfredoct96**
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | "features" 
| Bug fix?                               | no
| New feature?                           | yes
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | yes
| Related user documentation PR URL      | https://github.com/mautic/mautic-documentation/pull/155
| Related developer documentation PR URL | 
| Issue(s) addressed                     | 

### Integration Google Tag Manager on Mautic
In Setting Page -> Tracking Config i added a new panel with Google Tag Manager options.

Options includes:
* Google Tag Manager ID
* Enabled on your tracking page
* Enable on Mautic landing page

If ***Enabled on your tracking page*** is on, gtm is loaded on external page if user add mautic tracking code.

If ***Enable on Mautic landing page*** is on, gtm is loaded on all Mautic Landing Pages.

GTM script is loaded from backend, for external sites instead is loaded by mautic tracking code.

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Go to Settings -> Tracking Config
3. Edit Google Tag Manager fields, add a valid Google Tag Manager ID and enable the two checkboxes
4. Create a Landing Page
5. View the Landing Page and check if GTM is loaded

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/9493"></a>

### CODE REVIEW

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by RCheesley** [comment]
@alfredoct96 thanks for making this PR, sounds like a great feature! I have changed the base to 'features' as this is what we are using now for new features.

I have a couple of suggestions but will do that in a review :)

File: app/bundles/PageBundle/Translations/en_US/messages.ini
Comment (RCheesley) [inline app/bundles/PageBundle/Translations/en_US/messages.ini]: I would maybe make this a bit clearer what we mean by 'tracking page' to make it clear that we mean pages external to Mautic where the tracking code is being used:

```suggestion
mautic.page.config.form.tracking.trackingpage.enabled="Enabled on your externally tracked pages"
```

**Review by RCheesley — COMMENTED** [review]
I have made a small suggestion on the wording to make it clear when we are referring to externally tracked pages.

I think we might need to have a tooltip or help/warning text to remind people that they will need to check for double-loading of GTM and GA, for example, on external pages? 

Would we disable (or at least prompt to disable) the Google Analytics landing page settings if we had the GTM tracking enabled? Or show a warning about double-loading if there's already Google Analytics tags present?

I did not see a documentation PR (needs to be made against https://github.com/mautic/mautic-documentation please!) but we should make clear what will happen if that option is enabled, and GTM is already being loaded on the external pages?

You mentioned in the PR description that there are BC breaks, can you please list them out in the description?

**Comment by alfredoct96** [comment]
> I have made a small suggestion on the wording to make it clear when we are referring to externally tracked pages.
> 
> I think we might need to have a tooltip or help/warning text to remind people that they will need to check for double-loading of GTM and GA, for example, on external pages?
> 
> Would we disable (or at least prompt to disable) the Google Analytics landing page settings if we had the GTM tracking enabled? Or show a warning about double-loading if there's already Google Analytics tags present?
> 
> I did not see a documentation PR (needs to be made against https://github.com/mautic/mautic-documentation please!) but we should make clear what will happen if that option is enabled, and GTM is already being loaded on the external pages?
> 
> You mentioned in the PR description that there are BC breaks, can you please list them out in the description?

Hi @RCheesley , 
I have made some changes, I hope they go well .

1) Edited label text as you requested
2) When tracking page option is enabled on Google Analytics AND Google Tag Manager, or mautic landing page option is enabled on Google Analytics AND Google Tag Manager, i show a warning text,  recommending disabling one of the two options.
3) For BC Breaks is a my error, there aren't BC Breaks, i edited Pull request description.

I'm starting to write the documentation

**Comment by RCheesley** [comment]
@alfredoct96 I have had a bit of a struggle trying to rebase your PR - it seems that there is an issue with the branch, but also it is a protected organisation repo.

Can you please rebase against the Features branch?

Thanks!

**Comment by alfredoct96** [comment]
> @alfredoct96 I have had a bit of a struggle trying to rebase your PR - it seems that there is an issue with the branch, but also it is a protected organisation repo.
> 
> Can you please rebase against the Features branch?
> 
> Thanks!

Hi @RCheesley, sure!

I did the requested rebase.

**Comment by RCheesley** [comment]
@alfredoct96 it looks like there are a few files where we [need to improve test coverage](https://github.com/mautic/mautic/pull/9493#issuecomment-744508582) before we are able to merge this. Are you able to take a look at that?

**Comment by alfredoct96** [comment]
> @alfredoct96 it looks like there are a few files where we [need to improve test coverage](https://github.com/mautic/mautic/pull/9493#issuecomment-744508582) before we are able to merge this. Are you able to take a look at that?

@RCheesley  Sure, done!

**Comment by dennisameling** [comment]
Hi @alfredoct96! Thanks for your PR 🚀 we just upgraded Mautic to Symfony 4 and this PR will need to be updated. Since we don't have access to your branch, could you please click "Update branch" at the bottom of this PR? If that doesn't work, you can also run `git merge upstream/features` locally. Thanks in advance!

image

**Comment by alfredoct96** [comment]
> Hi @alfredoct96! Thanks for your PR  we just upgraded Mautic to Symfony 4 and this PR will need to be updated. Since we don't have access to your branch, could you please click "Update branch" at the bottom of this PR? If that doesn't work, you can also run `git merge upstream/features` locally. Thanks in advance!
> 
> image

Done @dennisameling

**Comment by mabumusa1** [comment]
@alfredoct96 Thanks for submitting this PR, do you have time to rebase it this weekend, I will test it directly once you rebase.

I think this is a great features to be shipped with Mautic 4

**Comment by alfredoct96** [comment]
> @alfredoct96 Thanks for submitting this PR, do you have time to rebase it this weekend, I will test it directly once you rebase.
> 
> I think this is a great features to be shipped with Mautic 4

@mabumusa1 Sure! I did the rebase

**Comment by nickveenhof** [comment]
Even though I think this is a great feature, I think this is more a configuration for in a plugin. Eg, Matomo Tag Manager also exists, and many others. How would we support those in the long run in Mautic? Maybe it can be a core plugin to load in any tag, but having specific support for Google Tag Manager only (instead of configurable/extendable) might be a bit too biased?

**Comment by swadeshp** [comment]
All tests has passed. Does it mean, it will be included in v4 this month?

**Comment by adiux** [comment]
I would like to see this in Mautic. Google Tag Manager is one (if not the) most used options to include tracking scripts. Maybe a bit small for a plugin? But I'm fine with any option.

File: app/bundles/PageBundle/Translations/en_US/messages.ini
Comment (RCheesley) [inline app/bundles/PageBundle/Translations/en_US/messages.ini]:
```suggestion
mautic.page.config.form.tracking.trackingpage.enabled="Enabled on your externally tracked pages"
```

File: app/bundles/PageBundle/Translations/en_US/messages.ini
Comment (RCheesley) [inline app/bundles/PageBundle/Translations/en_US/messages.ini]:
```suggestion
mautic.page.config.form.tracking.landingpage.enabled="Enable on Mautic landing pages"
```

**Review by RCheesley — CHANGES_REQUESTED** [review]
I think we should give information about the settings in the global config so I have left a comment on the docs for those - maybe could be useful as a tooltip too but will leave that to you.

I feel that we should also make it clear we are referring to externally tracked pages (not sure how this works - does it inject it into the tracking code somehow?) and landing pages in the plural.

Otherwise it works as expected and I was able to get my GTM code on the landing page and working as expected! 🚀

**Comment by RCheesley** [comment]
@mabumusa1 @adiux @swadeshp if we could get your tests on this PR I am hopeful we'll get it into 4.2 at the end of May - please take a look (you can test with Gitpod). Thanks!

**Review by RCheesley — DISMISSED** [review]
Re-approving based on my test [here](https://github.com/mautic/mautic/pull/9493#pullrequestreview-915464782) and changes being applied - thanks @alfredoct96 !

**Pull request closed without merging**
